flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/history/HistoryServer.java

```diff
@@ -0,0 +1,411 @@
+			// race, JVM is in shutdown already, we can safely ignore this
+			LOG.debug("Unable to add shutdown hook, shutdown already in progress", e);
+		} catch (Throwable t) {
+			// these errors usually happen when the shutdown is already in progress
+			LOG.warn("Error while adding shutdown hook", t);
+		}
+	}
+
+	public void run() {
+		try {
+			start();
+			while (true) {
+				try {
+					new CountDownLatch(1).await();
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+				}
+			}
+		} catch (Exception e) {
+			LOG.error("Failure while running HistoryServer.", e);
+		} finally {
+			stop();
+		}
+	}
+
+	// =================================================================================================================
+	// Life-cycle
+	// =================================================================================================================
+	private void start() throws IOException, InterruptedException {
+		synchronized (startupShutdownLock) {
+			LOG.info("Starting history server.");
+
+			Files.createDirectories(webDir.toPath());
+			LOG.info("Using directory {} as local cache.", webDir);
+			if (LOG.isInfoEnabled()) {
+				for (Path refreshDir : refreshDirs) {
+					LOG.info("Monitoring directory {} for archived jobs.", refreshDir);
+				}
+			}
+
+			Router router = new Router();
+			router.GET("/:*", new HistoryServerStaticFileServerHandler(webDir, new JobFileFetcher(new Path(webDir.getAbsolutePath()), cachedArchives)));
+
+			netty = new NettySetup(router, LOG, webDir, serverSSLContext, webAddress, webPort);
+
+			createDashboardConfigFile();
+			updateJobOverview(webDir);
+
+			executor.scheduleWithFixedDelay(fetcherTask, refreshIntervalMillis, refreshIntervalMillis, TimeUnit.MILLISECONDS);
+		}
+	}
+
+	private void stop() {
+		synchronized (startupShutdownLock) {
+			LOG.info("Stopping history server.");
+			netty.shutdown();
+			shutdownExecutor();
+			try {
+				LOG.info("Removing web dashboard root cache directory {}", webDir);
+				FileUtils.deleteDirectory(webDir);
+			} catch (Throwable t) {
+				LOG.warn("Error while deleting web root directory {}", webDir, t);
+			}
+
+			LOG.info("Stopped history server.");
+		}
+	}
+
+	private void shutdownExecutor() {
+		if (executor != null) {
+			executor.shutdown();
+
+			try {
+				if (!executor.awaitTermination(1, TimeUnit.SECONDS)) {
+					executor.shutdownNow();
+				}
+			} catch (InterruptedException ignored) {
+				executor.shutdownNow();
+			}
+		}
+	}
+
+	// =================================================================================================================
+	// File-fetching
+	// =================================================================================================================
+
+	/**
+	 * {@link TimerTask} that polls the directories configured as {@link HistoryServerOptions#HISTORY_SERVER_DIR} for
+	 * new job archives.
+	 */
+	private static class JobArchiveFetcherTask extends TimerTask {
```

Comment:
I would move the file fetch update logic to a separate outer class. The history server right now is very slim and essentially only a static file handler.